src/monte_carlo.c

```diff
@@ -0,0 +1,240 @@
+    if (game->isValidAction(board, &possibleActions[randomActionIdx],
+                            varPlayer)) {
+      // Play action
+      game->playAction(&simulationBoard, &possibleActions[randomActionIdx]);
+
+      // Remove action from possible actions
+      nPossibleActions -= simulationBoard.nPlayers;
+      game->removeAction(randomActionIdx, possibleActions, nPossibleActions);
+      int score = game->getScore(&simulationBoard, varPlayer);
+      if (score > game->drawValue) {
+        free(simulationBoard.values);
+        if (varPlayer == initialPlayer) {
+          return score;
+        } else {
+          return game->lossValue;
+        }
+      }
+      varPlayer = -(varPlayer);
+    }
+  }
+  free(simulationBoard.values);
+  return game->drawValue;
+}
+
+/**
+ * This method backpropagates the result of the playout to the root of the
+ * action tree.
+ * @param node The leaf node from which the backpropagation begins.
+ * @param score The score to backpropagate.
+ */
+void backpropagate(Node* node, int score) {
+  node->nVisits++;
+  node->score += score;
+
+  Node* currentChild = node;
+  Node* currentNode = node->parent;
+  while (currentNode) {
+    currentNode->nVisits += 1;
+    currentNode->score += score;
+    currentChild = currentNode;
+    currentNode = currentChild->parent;
+  }
+}
+
+/**
+ * This method frees the memory used during the Monte Carlo algorithm.
+ * @param node The root node from which the function starts freeing the memory.
+ */
+void freeMCTree(Node* node) {
+  for (unsigned i = 0; i < node->nChildren; ++i) {
+    freeMCTree(&node->children[i]);
+  }
+  free(node->children);
+  node->children = NULL;
+}
+
+/**
+ * This method executes the Monte Carlo algorithm.
+ * @param board The board used for the game.
+ * @param player The current player.
+ * @param game The definition of the game played, the environment in which the
+ * the machine learns.
+ * @param minSim Minimum number of simulations to run.
+ * @param maxSim Maximum number of simulations to run.
+ * @return The board with the best action to play after the execution of the
+ * Monte Carlo algorithm.
+ */
+Action mcGame(Board board, int player, Game game, int minSim, int maxSim,
+              mc_real goalValue) {
+  Node node = {.nVisits = 0,
+               .score = 0,
+               .action.player = -player,
+               .action.xPos = 0,
+               .action.yPos = 0,
+               .nChildren = 0,
+               .children = NULL,
+               .parent = NULL};
+
+  Board tempBoard;
+  tempBoard.values = malloc(game.getBoardSize() * sizeof(int8_t));
+  int iterations = 0;
+  while ((node.nVisits < minSim ||
+          calcUCB(findMaxUCB(node.children, node.nChildren)) < goalValue) &&
+         node.nVisits <= maxSim) {
+    for (int i = 0; i < game.getBoardSize(); ++i) {
+      tempBoard.values[i] = board.values[i];
+    }
+    tempBoard.nPlayers = board.nPlayers;
```

Comment:
You seem to make deep copies a few times (in this case with the tempBoard). Maybe make a function to deep copy a board